File: client/src/main/scala/stasis/client/ops/scheduling/OperationExecutor.scala
package stasis.client.ops.scheduling

import java.nio.file.Path
import java.time.Instant

import scala.concurrent.Future

import org.apache.pekko.Done

import stasis.client.collection.rules.RuleSet
import stasis.client.ops.recovery
import stasis.shared.model.datasets.DatasetDefinition
import stasis.shared.model.datasets.DatasetEntry
import stasis.shared.ops.Operation

trait OperationExecutor {

  def active: Future[Map[Operation.Id, Operation.Type]]
  def completed: Future[Map[Operation.Id, Operation.Type]]
  def rules: Future[RuleSet]

  def startBackupWithRules(
    definition: DatasetDefinition.Id
  ): Future[Operation.Id]

  def startBackupWithEntities(
    definition: DatasetDefinition.Id,
    entities: Seq[Path]
  ): Future[Operation.Id]

  def resumeBackup(
    operation: Operation.Id
  ): Future[Operation.Id]

  def startRecoveryWithDefinition(
    definition: DatasetDefinition.Id,
    until: Option[Instant],
    query: Option[recovery.Recovery.PathQuery],
    destination: Option[recovery.Recovery.Destination]
  ): Future[Operation.Id]

  def startRecoveryWithEntry(
    entry: DatasetEntry.Id,
    query: Option[recovery.Recovery.PathQuery],
    destination: Option[recovery.Recovery.Destination]
  ): Future[Operation.Id]

  def startExpiration(): Future[Operation.Id]

  def startValidation(): Future[Operation.Id]

  def startKeyRotation(): Future[Operation.Id]

  def stop(operation: Operation.Id): Future[Option[Done]]
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/ops/scheduling/OperationExecutor.kt
package stasis.client_android.lib.ops.scheduling

import stasis.client_android.lib.collection.rules.Rule
import stasis.client_android.lib.model.server.datasets.DatasetDefinitionId
import stasis.client_android.lib.model.server.datasets.DatasetEntryId
import stasis.client_android.lib.ops.Operation
import stasis.client_android.lib.ops.OperationId
import stasis.client_android.lib.ops.recovery.Recovery
import stasis.client_android.lib.ops.recovery.RecoverySourceKind
import java.nio.file.Path
import java.time.Instant

interface OperationExecutor {
    suspend fun active(): Map<OperationId, Operation.Type>
    suspend fun completed(): Map<OperationId, Operation.Type>
    suspend fun find(operation: OperationId): Operation.Type?

    suspend fun startBackupWithRules(
        definition: DatasetDefinitionId,
        rules: List<Rule>,
        f: (Throwable?) -> Unit
    ): OperationId

    suspend fun startBackupWithEntities(
        definition: DatasetDefinitionId,
        entities: List<Path>,
        f: (Throwable?) -> Unit
    ): OperationId

    suspend fun resumeBackup(
        operation: OperationId,
        f: (Throwable?) -> Unit
    ): OperationId

    suspend fun startRecoveryWithDefinition(
        definition: DatasetDefinitionId,
        until: Instant?,
        entities: Set<String>?,
        sources: Set<RecoverySourceKind>,
        destination: Recovery.Destination?,
        f: (Throwable?) -> Unit
    ): OperationId

    suspend fun startRecoveryWithEntry(
        entry: DatasetEntryId,
        entities: Set<String>?,
        sources: Set<RecoverySourceKind>,
        destination: Recovery.Destination?,
        f: (Throwable?) -> Unit
    ): OperationId

    suspend fun startExpiration(f: (Throwable?) -> Unit): OperationId

    suspend fun startValidation(f: (Throwable?) -> Unit): OperationId

    suspend fun startKeyRotation(f: (Throwable?) -> Unit): OperationId

    suspend fun stop(operation: OperationId)
}
